apps/src/templates/sectionProgress/progressTables/ProgressTableLevelBubble.jsx

```diff
@@ -3,181 +3,146 @@ import Radium from 'radium';
+    levelStatus: PropTypes.string,
+    levelKind: PropTypes.string,
+    isDisabled: PropTypes.bool,
+    isUnplugged: PropTypes.bool,
+    isConcept: PropTypes.bool,
+    isBonus: PropTypes.bool,
+    isPaired: PropTypes.bool,
+    bubbleSize: PropTypes.oneOf(Object.values(BubbleSize)).isRequired,
+    title: PropTypes.string,
+    url: PropTypes.string
   };
-}
 
-function largeContentStyle(props) {
-  const {bonus, paired} = props;
-  return {
-    ...progressStyles.flex,
-    ...styles.contents,
-    fontSize: paired || bonus ? 14 : 16
+  static defaultProps = {
+    bubbleSize: BubbleSize.full
   };
-}
 
-function UnpluggedBubble(props) {
-  return (
-    <div
-      style={{
-        ...styles.main,
-        ...styles.unplugged,
-        ...levelStyle(props)
-      }}
-    >
-      <Content {...props} />
-    </div>
-  );
-}
+  renderContent() {
+    const {
+      levelStatus,
+      isUnplugged,
+      isBonus,
+      isPaired,
+      title,
+      bubbleSize
+    } = this.props;
+    if (bubbleSize === BubbleSize.dot) {
+      // dot-sized bubbles are too small for content
+      return null;
+    }
+    const isLocked = levelStatus === LevelStatus.locked;
+    return isUnplugged ? (
+      <span>{i18n.unpluggedActivity()}</span>
+    ) : isLocked ? (
+      <FontAwesome icon="lock" />
+    ) : isPaired ? (
+      <FontAwesome icon="users" />
+    ) : isBonus ? (
+      <FontAwesome icon="flag-checkered" />
+    ) : title ? (
+      <span>{title}</span>
+    ) : null;
+  }
 
-function SmallCircle(props) {
-  return (
-    <div style={{...mainStyle(props), ...styles.smallCircle}}>
-      <Content {...props} />
-    </div>
-  );
+  render() {
+    const {isUnplugged, isConcept, isDisabled, bubbleSize} = this.props;
+    const bubbleShape = isUnplugged
+      ? BubbleShape.pill
+      : isConcept
+      ? BubbleShape.diamond
+      : BubbleShape.circle;
+
+    const bubble = (
+      <BasicBubble
+        shape={bubbleShape}
+        size={bubbleSize}
+        progressStyle={getProgressStyle(this.props)}
+      >
+        {this.renderContent()}
+      </BasicBubble>
+    );
+
+    if (isDisabled) {
+      return bubble;
+    }
+    return <LinkWrapper {...this.props}>{bubble}</LinkWrapper>;
+  }
 }
 
-function LargeBubble(props) {
-  const BubbleType = props.concept ? LargeDiamond : LargeCircle;
-  return (
-    <div style={styles.container}>
-      <BubbleType {...props} />
-    </div>
-  );
+/**
+ * Base bubble component defined in terms of shape, size, and style, as opposed
+ * to level-related properties. This component is itself only an empty styled
+ * container, and expects the bubble content (e.g. level number, icon, text) to
+ * be passed in as its children.
+ * @param {BubbleShape} shape
+ * @param {BubbleSize} size
+ * @param {object} progressStyle contains border and background colors
+ * @param {node} children the content to render inside the bubble
+ */
+function BasicBubble({shape, size, progressStyle, children}) {
```

Comment:
Would it make sense to make this its own component? If not, maybe we rename the function to `renderBasicBubble`?

Reply:
i want to defer that decision until I complete the rest of the refactoring work. for things like the table legends, it might be nice to have a bubble component that doesn't require a level object to render. and in general, having bottom-level components that are specified in terms of how they look instead of abstract concepts like "level" and "student progress" tends to allow for more reusability. 

i'm definitely open to putting this in its own module, but with how it's being used so far that didn't quite seem called for. depending on how much variation i find is needed to handle all our progress bubble cases, it might make sense to make this the primary `ProgressBubble` component and have a separate `BubbleFactory` module that computes shape, size, progressStyle, and content for different cases, with some helper wrappers for e.g. tooltips. once i get to that point i would love for us to discuss as a team since we don't have much of an existing pattern for that sort of thing.

even if we don't end up exporting this component for external use or moving it to its own module, i still think it's useful for testing that we're rendering the right shape/size for a given set of props. imo that pattern we've been playing around with feels cleaner than looking at pixel dimensions and geometry transforms, or mocking a `renderBasicBubble` function to see what it gets passed. but if you're not as into that pattern as i am i'm happy to discuss.

Reply:
I don't have a strong opinion right now about whether it needs to be its own component, so I'm fine if you want to leave it as is. I was thinking that moving it to it's own component would make this component cleaner, then we could also do individual tests for BasicBubble (which you could also do by exporting this). I'll leave it up to you!

Reply:
I like the bubble factory idea and am wondering if we want to encapsulate each bubble type more tightly by making each bubble type its own class and moving the style definitions, etc. for each type into the class.  There are some advantages to keeping the styles for all shapes/sizes in a single file, but I also like the idea of not having to look at code specific to diamonds when I'm trying to understand circles.